File: ch4/traversal/articulation.java
import java.util.*;
import java.io.*;

public class articulation {
  private static final int UNVISITED = 0;        // we only need this

  // these variables have to be global to be easily accessible by our recursion (other ways exist)
  private static ArrayList<ArrayList<IntegerPair>> AL;
  private static ArrayList<Integer> dfs_num, dfs_low, dfs_parent, articulation_vertex;
  private static int dfsNumberCounter, dfsRoot, rootChildren;

  private static void articulationPointAndBridge(int u) {
    dfs_low.set(u, dfsNumberCounter);
    dfs_num.set(u, dfsNumberCounter++); // dfs_low[u] <= dfs_num[u]
    for (IntegerPair v_w : AL.get(u)) {
      if (dfs_num.get(v_w.first()) == UNVISITED) { // a tree edge
        dfs_parent.set(v_w.first(), u);
        if (u == dfsRoot) ++rootChildren;        // special case, root

        articulationPointAndBridge(v_w.first());

        if (dfs_low.get(v_w.first()) >= dfs_num.get(u)) // for articulation point
          articulation_vertex.set(u, 1); // store this information first
        if (dfs_low.get(v_w.first()) > dfs_num.get(u)) // for bridge
          System.out.printf(" Edge (%d, %d) is a bridge\n", u, v_w.first());
        dfs_low.set(u, Math.min(dfs_low.get(u), dfs_low.get(v_w.first()))); // update dfs_low[u]
      }
      else if (v_w.first() != dfs_parent.get(u)) // a back edge and not direct cycle
        dfs_low.set(u, Math.min(dfs_low.get(u), dfs_num.get(v_w.first()))); // update dfs_low[u]
    }
  }

  public static void main(String[] args) throws Exception {
    /*
    // Left graph in Figure 4.6/4.7/4.8
    6
    1 1 0
    3 0 0 2 0 4 0
    1 1 0
    1 4 0
    3 1 0 3 0 5 0
    1 4 0

    // Right graph in Figure 4.6/4.7/4.8
    6
    1 1 0
    5 0 0 2 0 3 0 4 0 5 0
    1 1 0
    1 1 0
    2 1 0 5 0
    2 1 0 4 0
    */

    Scanner sc = new Scanner(new File("articulation_in.txt"));

    int V = sc.nextInt();
    AL = new ArrayList<>();
    for (int u = 0; u < V; ++u) {
      AL.add(new ArrayList<>());                 // store blank vector first
      int k = sc.nextInt();
      while (k-- > 0) {
        int v = sc.nextInt(), w = sc.nextInt();
        AL.get(u).add(new IntegerPair(v, w));
      }
    }

    System.out.printf("Articulation Points & Bridges (the input graph must be UNDIRECTED)\n");
    dfs_num = new ArrayList<>(Collections.nCopies(V, UNVISITED)); // global variable
    dfs_low = new ArrayList<>(Collections.nCopies(V, 0));
    dfs_parent = new ArrayList<>(Collections.nCopies(V, -1));
    articulation_vertex = new ArrayList<>(Collections.nCopies(V, 0));
    dfsNumberCounter = 0;
    System.out.printf("Bridges:\n");
    for (int u = 0; u < V; ++u)
      if (dfs_num.get(u) == UNVISITED) {
        dfsRoot = u; rootChildren = 0;
        articulationPointAndBridge(u);
        articulation_vertex.set(dfsRoot, (rootChildren > 1) ? 1 : 0); // special case
      }

    System.out.printf("Articulation Points:\n");
    for (int u = 0; u < V; ++u)
      if (articulation_vertex.get(u) == 1)
        System.out.printf(" Vertex %d\n", u);
  }
}

File: ch4/traversal/articulation.ml
open Scanf
open Printf

type graph = int list array (* Adjacency List representation *)

let range n = List.init n (fun x -> x)

(** Returns a list of all articulation points and a list of all bridges of the
    simple graph.
    For multigraphs, store edge ids to the adjacency lists to check for
    bidirectional edges instead of using [parent] *)
let articulation_point_and_bridge (adj : graph) : int list * (int*int) list =
  let n = Array.length adj in
  let low = Array.make n 0 in
  let num = Array.make n (-1) in
  let parent = Array.make n (-1) in
  let counter = ref 0 in
  let rec visit u =
    low.(u) <- !counter;
    num.(u) <- !counter;
    incr counter;
    adj.(u) |> List.iter (fun v ->
      if num.(v) = -1 then begin
        parent.(v) <- u;
        visit v;
        low.(u) <- min low.(u) low.(v)
      end
      else if v <> parent.(u) then (** Use edge id for multigraphs *)
        low.(u) <- min low.(u) num.(v)
    )
  in
  range n |> List.iter (fun u ->
    if num.(u) = -1 then begin
      counter := 0;
      visit u
    end
  );
  let articulation_points = range n |> List.filter (fun u ->
    if num.(u) = 0 then (* special case for root *)
      adj.(u) |> List.filter (fun v -> parent.(v) = u) |> List.length > 1
    else adj.(u) |> List.exists (fun v -> parent.(v) = u && low.(v) >= num.(u))
  ) in
  let bridges = range n |> List.map (fun u ->
    adj.(u)
      |> List.filter (fun v -> parent.(v) = u && low.(v) > num.(u))
      |> List.map (fun v -> (u, v))
  ) |> List.concat in
  articulation_points, bridges

let () =
  let ic = Scanning.open_in "articulation_in.txt" in
  let sc = object method nf = bscanf ic end in
  let n = sc#nf " %d " (fun x -> x) in
  let adj = Array.init n (fun _ ->
    sc#nf " %d " (fun degree ->
      List.init degree (fun _ -> sc#nf " %d %d " (fun v w -> v))) (* Ignoring weight *)
  ) in
  printf "Articulation Points & Bridges (the input graph must be UNDIRECTED)\n";
  let aps, bridges = articulation_point_and_bridge adj in
  printf "Bridges:\n";
  bridges |> List.iter (fun (u, v) -> printf " Edge (%d, %d) is a bridge\n" u v);
  printf "Articulation Points:\n";
  aps |> List.iter (printf " Vertex %d\n");
